src/renderers/shared/fiber/ReactFiberUpdateQueue.js

```diff
@@ -12,100 +12,473 @@
+    partialState: null,
     callback: null,
     isReplace: false,
+    isForced: true,
     next: null,
   };
-  queue.tail.next = node;
-  queue.tail = node;
-  queue.hasUpdate = queue.hasUpdate || (partialState != null);
-  return queue;
+  if (__DEV__) {
+    insertUpdate(fiber, update, 'forceUpdate');
+  } else {
+    insertUpdate(fiber, update);
+  }
 }
+exports.addForceUpdate = addForceUpdate;
 
-exports.addToQueue = addToQueue;
 
-exports.addCallbackToQueue = function(queue : UpdateQueue, callback: Function) : UpdateQueue {
-  if (queue.tail.callback) {
-    // If the tail already as a callback, add an empty node to queue
-    addToQueue(queue, null);
+function addCallback(fiber : Fiber, callback: Callback, priorityLevel : PriorityLevel) : void {
+  const update : Update = {
+    priorityLevel,
+    partialState: null,
+    callback,
+    isReplace: false,
+    isForced: false,
+    next: null,
+  };
+  insertUpdate(fiber, update);
+}
+exports.addCallback = addCallback;
+
+function getPendingPriority(queue : UpdateQueue) : PriorityLevel {
+  return queue.first ? queue.first.priorityLevel : NoWork;
+}
+exports.getPendingPriority = getPendingPriority;
+
+function getStateFromUpdate(update, instance, prevState, props) {
+  const partialState = update.partialState;
+  if (typeof partialState === 'function') {
+    const updateFn = partialState;
+    return updateFn.call(instance, prevState, props);
+  } else {
+    return partialState;
   }
-  queue.tail.callback = callback;
-  queue.hasCallback = true;
-  return queue;
-};
+}
 
-exports.callCallbacks = function(queue : UpdateQueue, context : any) {
-  let node : ?UpdateQueueNode = queue;
-  while (node) {
-    const callback = node.callback;
-    if (callback) {
-      if (typeof context !== 'undefined') {
-        callback.call(context);
+function beginUpdateQueue(
+  workInProgress : Fiber,
+  queue : UpdateQueue,
+  instance : any,
+  prevState : any,
+  props : any,
+  priorityLevel : PriorityLevel
+) : any {
+  if (__DEV__) {
+    // Set this flag so we can warn if setState is called inside the update
+    // function of another setState.
+    queue.isProcessing = true;
+  }
+
+  // Applies updates with matching priority to the previous state to create
+  // a new state object.
+  let state = prevState;
+  let dontMutatePrevState = true;
+  let isEmpty = true;
+  let callbackList = null;
+  let update = queue.first;
+  while (update && comparePriority(update.priorityLevel, priorityLevel) <= 0) {
+    // Remove each update from the queue right before it is processed. That way
+    // if setState is called from inside an updater function, the new update
+    // will be inserted in the correct position.
+    queue.first = update.next;
+    if (!queue.first) {
+      queue.last = null;
+    }
+
+    let partialState;
+    if (update.isReplace) {
+      // A replace should drop all previous updates in the queue, so
+      // use the original `prevState`, not the accumulated `state`
```

Comment:
Naming is bad. `state` is the accumulated state. `prevState` is the initial accumulation.

Roughly what this code is doing is:

```js
// initialState is the state before applying any updates
const nextState = updates.reduce((accumulatedState, update) => {
  if (update.isReplace) {
    return merge(initialState, update.partialState);
  } else {
    return merge(accumulatedState, update.partialState);
  }
}, initialState);
```

By applying the update to the initial state rather than the accumulation, it effectively drops the previous updates. But... this actually isn't right anyway because we should only drop updates with the same priority. We already do this in `addReplaceState`. So this comment is wrong and we should always pass `state` (the accumulation).